### PR TITLE
fix(BUX-551): redundant methods and webhooks after duplicated block instert

### PR DESCRIPTION
**Pull request by wregulski**
# Pull Request Checklist

- [x] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/bitcoin-sv/pulse/blob/main/.github/CODE_STANDARDS.md)
- [x] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/pulse/blob/main/.github/CODE_OF_CONDUCT.md)
- [x] 🏠 I tested my changes locally.
- [ ] ✅ I have provided tests for my changes.
- [x] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [x] 💾 PR was issued based on the Github or Jira issue.

### CODE REVIEW

File: errors/unique_violation.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package errors
+
+type UniqueViolationError struct {
```
Comment (wregulski) [inline errors/unique_violation.go:3]: If you, my dear reviewer, have another idea on how to pass such an error between layers in order not to use "magic strings" then I will be very grateful for your suggestion :)

File: database/sql/headers.go [hunk 2]
```diff
@@ -187,11 +195,30 @@ func (h *HeadersDb) Create(ctx context.Context, req dto.DbBlockHeader) error {
...
+		}
+
 		return errors.Wrap(err, "failed to insert header")
 	}
+
 	return errors.Wrap(tx.Commit(), "failed to commit tx")
 }
 
+func (h *HeadersDb) isUniqueViolation(err error) bool {
+	switch h.db.DriverName() {
```
Comment (chris-4chain) [inline database/sql/headers.go:209]: Here I'm thinking about type switch between `sqlite3.Error` and `pq.Error`. In this approach, the DriverName would not be necessary.

File: service/chain_service.go [hunk 3]
```diff
@@ -188,6 +189,11 @@ func (cs *chainService) previousHeader(bs *domains.BlockHeaderSource) (*domains.
 func (cs *chainService) insert(h *domains.BlockHeader) (*domains.BlockHeader, error) {
 	err := cs.Repositories.Headers.AddHeaderToDatabase(*h)
 	if err != nil {
+		if errors.Is(err, customErrs.NewUniqueViolationError()) {
```
Comment (arkadiuszos4chain) [inline service/chain_service.go:192]: I don't understand how this code would prevent multiple webhook calls after `insert()`

Comment (wregulski) [reply service/chain_service.go:196]: fixed

File: database/sql/headers.go (hunk 2 above)
Comment (wregulski) [reply database/sql/headers.go]: fixed - replaced switch DriverName with switch err(type)

File: service/chain_service.go (hunk 3 above)
Comment (arkadiuszos4chain) [reply service/chain_service.go:196]: I believe we don't need this check here. Please move warning to Add() method

File: database/sql/headers.go [hunk 4]
```diff
@@ -187,11 +193,27 @@ func (h *HeadersDb) Create(ctx context.Context, req dto.DbBlockHeader) error {
 		_ = tx.Rollback()
 	}()
 	if _, err := tx.NamedExecContext(ctx, sqlInsertHeader, req); err != nil {
+		if isUniqueViolation(err) {
```
Comment (arkadiuszos4chain) [inline database/sql/headers.go:196]: Are you sure we will get this error here, not on `tx.Commit()` ?

**Review by arkadiuszos4chain — CHANGES_REQUESTED** [review]

File: database/sql/headers.go (hunk 4 above)
Comment (wregulski) [reply database/sql/headers.go:196]: Yes, 100% sure

File: service/chain_service.go (hunk 3 above)
Comment (wregulski) [reply service/chain_service.go:196]: Moved to Add method

**Review by arkadiuszos4chain — APPROVED** [review]

**Review by chris-4chain — APPROVED** [review]

**Pull request merged**